### PR TITLE
skeleton for dataloader, needs linked in cmake for all my cpp files

### PR DESCRIPTION
**Pull request by csmaster23**

### CODE REVIEW

File: include/load_data.hpp [hunk 1]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+#include <string>
+#include <vector>
+
+#include"opencv2\opencv.hpp"
+// #include "opencv2/opencv.hpp"
```
Comment (taylorpool) [inline include/load_data.hpp]: Can you remove these lines?

File: src/slamdunk.cpp [hunk 2]
```diff
@@ -0,0 +1,20 @@
+#include <iostream>
+#include <string>
+#include "../include/load_data.hpp"
+#include "./load_data.cpp"
+
+int main()
+{
+    std::cout << "--- SLAM-DUNK ---" << std::endl;
+
+    // probably want a better place to put these but for now
```
Comment (taylorpool) [inline src/slamdunk.cpp:9]: What are your thoughts about creating a .json file for this stuff?

File: include/load_data.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// #include "opencv2/highgui/highgui.hpp"
+
+class Dataloader
+{
+public:
+  std::string dataset_name;
+  std::string dataset_path;
+
+  std::vector<cv::Mat> mat_images;
```
Comment (taylorpool) [inline include/load_data.hpp]: Could we get a more descriptive variable name here?

**Review by taylorpool — COMMENTED** [review]
See comments

**Comment by csmaster23** [comment]
New pull request

File: include/load_data.hpp [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+#include "opencv2/opencv.hpp"
+#include <opencv2/features2d.hpp>
+
+class Dataloader
+{
+public:
+  std::string dataset_name;
+  std::string dataset_path;
+
+  std::vector<std::vector<cv::KeyPoint>> feature_vec;
```
Comment (taylorpool) [inline include/load_data.hpp:22]: Can you explain a little more about what's going on here?

File: include/load_data.hpp [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+  // Constructor
+  Dataloader(std::string name_in, std::string path_in);
+
+  // Creates the image vector as Mat opencv datatypes
+  void create_feature_vectors();
+
+  // Helper Functions
+  std::string create_image_title(int counter);
+  void save_image_with_features(cv::Mat image_in, std::vector<cv::KeyPoint> key_points, std::string save_path);
```
Comment (taylorpool) [inline include/load_data.hpp:54]: Can you please add documentation for each of these functions? i.e. purpose, inputs, outputs, etc

File: src/load_data.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+void Dataloader::create_feature_vectors() {
+    // Delete this after our two images test
+    int counter = 0;
+    bool plot = false;
+
+    cv::Ptr<cv::FeatureDetector> detector = cv::ORB::create();
+    
+    for (const auto & entry : std::filesystem::directory_iterator(dataset_path)) 
+    {
+        // --- Delete this after our two images test ---
```
Comment (taylorpool) [inline src/load_data.cpp]: Mark this as TODO

File: src/load_data.cpp [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+        if (counter > 1) {
+            break;
+        }
+        // --- Delete this after our two images test ---
+        
+        const cv::Mat image = cv::imread(entry.path(), 0); //Load as grayscale
```
Comment (taylorpool) [inline src/load_data.cpp:34]: Can you put all comments on top of the code? Just for consistency's sake

File: src/load_data.cpp [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+        std::vector<cv::KeyPoint> keypoints;
+        detector->detect(image, keypoints);
+
+        feature_vec.push_back( keypoints );  // add keypoints to keypoint vec
+
+        // Add results to image and save.
+        if (plot)
```
Comment (taylorpool) [inline src/load_data.cpp]: Can you put all of this plotting logic in a separate function? Ideally we should try to keep functions no more than 20 lines or so

**Review by taylorpool — CHANGES_REQUESTED** [review]
Reviewed

File: include/load_data.hpp [hunk 9]
```diff
@@ -8,9 +8,17 @@
 class Dataloader
 {
 public:
+  // name of the dataset we want to use
   std::string dataset_name;
```
Comment (taylorpool) [inline include/load_data.hpp:12]: If we have getters and setters for these fields, then these should be private and not public.

**Review by taylorpool — CHANGES_REQUESTED** [review]

**Review by taylorpool — APPROVED** [review]

**Comment by taylorpool** [comment]
Nice work

**Pull request merged**